operations/research-and-development/product/development-process/mobile-feature-guidelines.md

```diff
@@ -0,0 +1,39 @@
+    - Example: Users cannot see the “Mark as Unread” option in an older mobile app that does not contain that code, but marking a channel unread on a webapp running v5.20 or later should still mark the channel as unread on mobile, and not cause any mobile app regressions for a user running an old app. 
+  - Ideally any new features should be tested on all server versions back to the [oldest supported ESR](https://docs.mattermost.com/administration/extended-support-release.html?highlight=esr). However, since testing mana is limited, feel free to work with your QA representative to determine the most effective testing approach based on the risks of your feature. At a minimum, new mobile app features should be tested on all supported server ESRs and the latest three server versions.
+- For new features, consider a minserverversion check that only executes the code if the server connected to the app is newer than xyz. [See example to disable or not the position field when editing the profile](https://github.com/mattermost/mattermost-mobile/blob/ee4b85edcfee8316db08c31ec5b2a26afb343bd3/app/screens/edit_profile/index.js#L29)
+Bonus: Legacy servers (ie <v5.0) don't have post metadata, don't rely on it.
+
+## #3 Variable Network Conditions
+
+- API requests will fail on mobile because network conditions are much more variable than on desktop. As such defaults need to be carefully chosen to avoid failing requests breaking or blocking core user functionality. Example: if the default for a channel is set to be read only until a permission API request grants the user permissions, this is likely to result in a poor user experience in bad networks. When in doubt, [the default if an API request fails should not change existing behaviour, or a failed API request should either do nothing or notify the user somehow](https://github.com/mattermost/mattermost-mobile/blob/master/app/mm-redux/actions/preferences.ts#L18)
+- [Retries should be added to important API requests](https://github.com/mattermost/mattermost-mobile/blob/master/app/actions/views/channel.js#L607)
+
+## #4 Client Performance Impact
+
+- Mobile devices vary significantly in their computing capabilities, as such performance should be top of mind when developing mobile features.
+- Minimize requests where possible, make requests in parallel if possible, batch dispatches wherever possible. 
+https://github.com/mattermost/mattermost-mobile/blob/master/app/actions/views/user.js#L79
+- Do NOT track requests. Examples:
+  - Here we are fetching the needed data in sequence but holding dispatching any action, and then once we have all the required data we dispatch an action batch. With this we ensure that the redux store is being updated only once and that every mapStateToProps of mounted connected components runs once instead of multiple times, this will then reduce computation time: https://github.com/mattermost/mattermost-mobile/blob/master/app/actions/views/user.js#L51
```

Comment:
```suggestion
  - Here we are fetching the needed data in sequence but holding dispatching any action, and then once we have all the required data we dispatch an action batch. With this we ensure that the redux store is being updated only once and that every `mapStateToProps` of mounted connected components runs once instead of multiple times, this will then reduce computation time: https://github.com/mattermost/mattermost-mobile/blob/master/app/actions/views/user.js#L51.
```